sql: Cancel txn context on query cancellation

itsbilal opened the pull request:
This reverts some of the changes made in #17003 and #17200. In particular:

#17003 made new contexts for every statement, but didn't cancel them after statement execution, resulting in a memory leak. 
#17200 did a quick-fix for that memory leak by not forking that context, at the expense of not being able to leverage the context in cancellation.

This PR restores the context cancellation behaviour introduced in #17003 , except at the transaction level - cancelling a query closes its transaction's context. The `TxnCoordSender` [isn't set up to handle contexts with a smaller lifetime than the transaction itself](https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/txn_coord_sender.go#L619),

Commit:
sql: Cancel txn context on query cancellation.

This reverts some of the changes made in PRs #17003 and #17200. In
particular:

17003 made new contexts for every statement, but didn't cancel them
after statement execution, resulting in a memory leak.
17200 did a quick-fix for that memory leak by not forking that context,
at the expense of not being able to leverage the context in
cancellation.

This PR restores the context cancellation behaviour introduced in #17003
, except at the transaction level - cancelling a query closes its
transaction's context.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17219)
<!-- Reviewable:end -->

tamird commented:
Can you please include all this detail in the *commit* message?

itsbilal commented:
Added the same details to the commit message.

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/query_cancellation.md, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzHifCSYnLTK0RJ9Is:-KpzpV6X8Lj1-VfFwHDf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L106)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please always use links to a specific revision, otherwise the line number will go stale real quick. You can press `y` in github to get a "canonical url".
</blockquote></details>

Done.

---

*[docs/RFCS/query_cancellation.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzIguHJLQO9nDQLS7r:-KpzpPp-LQTP7GNSe9rL:biss4x5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L107)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

but what about parallel queries? If we're doing this flag instead of checking the context, then we should go through the parallelizeQueue and flag all the running queries.
</blockquote></details>

The `cancelChecker` actually checks for both the flag and the context (but only once every 100k rows), so explicitly setting that flag would be unnecessary. That way, we can distinguish between cancellations directly intended for a query, from all other cancellations. 

---

*[docs/RFCS/query_cancellation.md, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzKVzMOIwQlb7jPHbz:-Kpzoq8MDl2o1O4Mtfmv:b-esf5r1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L124)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

did you actually test this? How did it look?
</blockquote></details>

I timed both approaches around the time I wrote the RFC, and it was about ~20x slower to check a `context.WithCancel(context.Background())` than it was to load a shared int atomically. The overhead would only go up for contexts that are several derivations away from `context.Background()`, as would be the case in practice.

---

*[pkg/sql/executor.go, line 372 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzLHdh49_aP-UXdZDQ:-KpzyO-n0f8ivG2-8sa1:b8fsop1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/pkg/sql/executor.go#L372)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can it be nil? I hope not, but if it can, please comment about what that means on the field declaration.
</blockquote></details>

Done. You're right - the `queryMeta` one can't be nil, but the one in `txnState` can be (when state = `NoTxn`). Clarifying.

---

*[pkg/sql/executor.go, line 377 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzLkEiOpwAKZxd5h4b:-KpzyF3HKEfc7r7ddzEZ:bkopdl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/pkg/sql/executor.go#L377)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this should be part of the `cancel()` callback
</blockquote></details>

`cancel()` is just a context `CancelFunc`, not an actual method on `queryMeta`. And @israelg99 has an upcoming PR where he moves this to a `Cancel()` method on `queryMeta` - but for now this is closer to the #17003 so I'd rather stick to that to reduce conflicts.

---

*[pkg/sql/executor.go, line 1110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzM45X_nCA7nCnVh7v:-Kq-0hrD6fDvaRIt1kkp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/pkg/sql/executor.go#L1110)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment here saying that we can't fork off contexts for individual queries cause the txncoordsender captures the ctx of the first op as the txn ctx
</blockquote></details>

Done.

---

*[pkg/sql/session.go, line 1005 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzLIP8gHY4B_gS1J8m:-Kq-0idH94UAXOKSiIqo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/pkg/sql/session.go#L1005)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ditto about nil
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-:-Kq-791oaJsNRUpkV9uY:b-9t66dc)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/query_cancellation.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzIguHJLQO9nDQLS7r:-Kq-EtEfXf5ZsDCtI8x9:bxtw4f8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L107)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

The `cancelChecker` actually checks for both the flag and the context (but only once every 100k rows), so explicitly setting that flag would be unnecessary. That way, we can distinguish between cancellations directly intended for a query, from all other cancellations. 
</blockquote></details>

either there's a difference between cancelling  a query and "all other cancellations" or there isn't. This PR makes generally it so that there isn't. So the argument that you can make a distinction between the two doesn't hold, I think.
I think the flag and the context cancellation should always be in sync, and we shouldn't check both (except as an assertion).

---

*[pkg/sql/executor.go, line 377 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzLkEiOpwAKZxd5h4b:-Kq-FkI8zq5qrhzyxN61:b-azq2o1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/pkg/sql/executor.go#L377)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

`cancel()` is just a context `CancelFunc`, not an actual method on `queryMeta`. And @israelg99 has an upcoming PR where he moves this to a `Cancel()` method on `queryMeta` - but for now this is closer to the #17003 so I'd rather stick to that to reduce conflicts.
</blockquote></details>

don't know what that other PR does, but IMHO this is the time to bundle setting the flag and cancelling the context. From the perspective of this code, the fact that currently `queryMeta.cancel` is directly taken from what `context.WithCancel()` returns is irrelevant. All it knows is that, to cancel a query, you need to 1) call `cancel()` and 2) set `isCancelled=1`. One too many things to do :) 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/query_cancellation.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzIguHJLQO9nDQLS7r:-Kq-KtUMMp0TDzFycwhj:bfrpo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L107)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

either there's a difference between cancelling  a query and "all other cancellations" or there isn't. This PR makes generally it so that there isn't. So the argument that you can make a distinction between the two doesn't hold, I think.
I think the flag and the context cancellation should always be in sync, and we shouldn't check both (except as an assertion).
</blockquote></details>

Maybe this was too strong; I guess cancelling a txn context can be observed by parallel queries before we have a chance to set their `isCancelled` flags no matter what we do, so the two can't be perfectly in sync. So I've relaxed my stance, I no longer insist on going through the parallelizeQueue and informing everybody explicitly, although I still think it'd be cool.
But this fact that you have to check both has to be explained somewhere - consider moving the checks from `cancelChecker` to a `queryMeta.IsQueryCancelled()` method, and explain why we check both.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/query_cancellation.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzIguHJLQO9nDQLS7r:-Kq-Mo6seTM0sPmSVcVe:b-62pw8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L107)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Maybe this was too strong; I guess cancelling a txn context can be observed by parallel queries before we have a chance to set their `isCancelled` flags no matter what we do, so the two can't be perfectly in sync. So I've relaxed my stance, I no longer insist on going through the parallelizeQueue and informing everybody explicitly, although I still think it'd be cool.
But this fact that you have to check both has to be explained somewhere - consider moving the checks from `cancelChecker` to a `queryMeta.IsQueryCancelled()` method, and explain why we check both.
</blockquote></details>

Or, the `queryMeta.cancel()` method can use a structure that accumulates all queries that share the context and sets all their flags before calling the context's cancel method
:sunglasses: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

itsbilal commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/query_cancellation.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzIguHJLQO9nDQLS7r:-Kq-iwIyT0i_if2NbqPc:b-8l6u85) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L107)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Or, the `queryMeta.cancel()` method can use a structure that accumulates all queries that share the context and sets all their flags before calling the context's cancel method
:sunglasses: 
</blockquote></details>

I thought about this a bit more - and now I'm leaning towards not checking the context for cancellation in `isQueryCancelled` (which I've added as a method.).  It made a lot more sense when we had per-statement contexts back in #17003, but now that we don't, we should just stick to checking the cancellation flag in planNodes. I'm still cancelling the txn context when I cancel the query, though.

From a user-facing perspective, it means that the entire transaction goes toast (which it should), but only the intended target query gets the 'query execution cancelled' error.

---

*[pkg/sql/executor.go, line 377 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzLkEiOpwAKZxd5h4b:-Kq-YUpKgRsrdeDgDJkF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/pkg/sql/executor.go#L377)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

don't know what that other PR does, but IMHO this is the time to bundle setting the flag and cancelling the context. From the perspective of this code, the fact that currently `queryMeta.cancel` is directly taken from what `context.WithCancel()` returns is irrelevant. All it knows is that, to cancel a query, you need to 1) call `cancel()` and 2) set `isCancelled=1`. One too many things to do :) 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/query_cancellation.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzIguHJLQO9nDQLS7r:-Kq-ri6HPMfOMRRJnBSt:b-jzk6tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L107)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I thought about this a bit more - and now I'm leaning towards not checking the context for cancellation in `isQueryCancelled` (which I've added as a method.).  It made a lot more sense when we had per-statement contexts back in #17003, but now that we don't, we should just stick to checking the cancellation flag in planNodes. I'm still cancelling the txn context when I cancel the query, though.

From a user-facing perspective, it means that the entire transaction goes toast (which it should), but only the intended target query gets the 'query execution cancelled' error.
</blockquote></details>

Well but you want to want to stop the other queries as soon as possible. One of the points of cancelling a query is that you want the client that sent it to get some error quickly, but if that client is blocked on multiple parallel queries, this won't happen.

It's true that, if the txn context is cancelled, the `TxnCoordSender` will find out and abandon the transaction, and then every other query that tries to send requests through it will get an ugly "no state" error, but I don't think we should rely on that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

itsbilal commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/query_cancellation.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KpzIguHJLQO9nDQLS7r:-Kq3Un4pjXhQHw7_3ldE:bdae8ve) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e47d05302345c946d0606050cfceaf2048d7e7f/docs/RFCS/query_cancellation.md#L107)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Well but you want to want to stop the other queries as soon as possible. One of the points of cancelling a query is that you want the client that sent it to get some error quickly, but if that client is blocked on multiple parallel queries, this won't happen.

It's true that, if the txn context is cancelled, the `TxnCoordSender` will find out and abandon the transaction, and then every other query that tries to send requests through it will get an ugly "no state" error, but I don't think we should rely on that.
</blockquote></details>

Good point. Added the context check back to `isQueryCancelled`, with a comment explaining why.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but a test for the cancellation of a parallel query would be nice. You can use a statement filter to block a query, then cancel another one in its txn, then unblock it and see that it returns an error. This is assuming that we query cancellation currently really does something; I lost track of whether that's true or not.

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 202 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq3fk5AIA3Um5g-LQMC:-Kq3fk5AIA3Um5g-LQMD:bmflzxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/02472b8cbf708fb592c7eddd4de1ea289fba4bf8/pkg/sql/session.go#L202)):*
> ```Go
> 	q.ctxCancel()
> 
> 	// Atomically set isCancelled to 1
> ```

nit: this comment is unnecessary; you can just read the next line :)

---

*[pkg/sql/session.go, line 213 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq3fzCyVVj0QDwCu2p7:-Kq3fzCyVVj0QDwCu2p8:b-xhe7i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/02472b8cbf708fb592c7eddd4de1ea289fba4bf8/pkg/sql/session.go#L213)):*
> ```Go
> // transaction context will be closed but isCancelled won't be set to true.
> func (q *queryMeta) isQueryCancelled() bool {
> 	// Atomically check shared cancellation flag first.
> ```

ditto nit - unnecessary comment. If the emphasis is on "first" and you think it's important, then consider expanding that (i.e. "because it's cheaper").

---

*[pkg/sql/session.go, line 842 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq3hQjveT-8x-7XfWAR:-Kq3hQjveT-8x-7XfWAS:b-sg48p1) ([raw file](https://github.com/cockroachdb/cockroach/blob/02472b8cbf708fb592c7eddd4de1ea289fba4bf8/pkg/sql/session.go#L842)):*
> ```Go
> 	// cancel is the cancellation function for the above context. Called upon
> 	// COMMIT/ROLLBACK of the transaction to release resources associated with
> 	// the context. Can be nil when no txn is in progress.
> ```

s/can be/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-:-Kq3hY1Bl5UTG8CBPUFx:b-hl91b3)*
<!-- Sent from Reviewable.io -->

itsbilal commented:
Added a test for the parallel case.

---

Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/session.go, line 202 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq3fk5AIA3Um5g-LQMC:-Kq4RvwJ-r5Q0IwtReuH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02472b8cbf708fb592c7eddd4de1ea289fba4bf8/pkg/sql/session.go#L202)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: this comment is unnecessary; you can just read the next line :)
</blockquote></details>

Done.

---

*[pkg/sql/session.go, line 213 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq3fzCyVVj0QDwCu2p7:-Kq4RwGET0GS6dI-v8CK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02472b8cbf708fb592c7eddd4de1ea289fba4bf8/pkg/sql/session.go#L213)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ditto nit - unnecessary comment. If the emphasis is on "first" and you think it's important, then consider expanding that (i.e. "because it's cheaper").
</blockquote></details>

Done.

---

*[pkg/sql/session.go, line 842 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq3hQjveT-8x-7XfWAR:-Kq4RyDZRemINQofd5fs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02472b8cbf708fb592c7eddd4de1ea289fba4bf8/pkg/sql/session.go#L842)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/can be/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-:-Kq4RK-3TgPaB4NeSS-p:b3nf3bp)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/session.go, line 213 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq3fzCyVVj0QDwCu2p7:-Kq4VfD4D4vwee4u3aYy:b-q9xcza) ([raw file](https://github.com/cockroachdb/cockroach/blob/02472b8cbf708fb592c7eddd4de1ea289fba4bf8/pkg/sql/session.go#L213)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

hold up, I'm confused. So the point of `q.isCancelled` is that it's cheaper to check than the ctx. But in the common case (not cancelled), we check both - so what's the point?. I know I told you to check both :), but now this seems wrong.

If we keep the flag, we should check the ctx less often.
Or, consider again the suggestion to set all flags for all parallel queries in sync, and stop checking the context.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

itsbilal commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/session.go, line 213 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq3fzCyVVj0QDwCu2p7:-Kq4cQBKAgXsRCnMFUCK:b-wxwev2) ([raw file](https://github.com/cockroachdb/cockroach/blob/02472b8cbf708fb592c7eddd4de1ea289fba4bf8/pkg/sql/session.go#L213)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

hold up, I'm confused. So the point of `q.isCancelled` is that it's cheaper to check than the ctx. But in the common case (not cancelled), we check both - so what's the point?. I know I told you to check both :), but now this seems wrong.

If we keep the flag, we should check the ctx less often.
Or, consider again the suggestion to set all flags for all parallel queries in sync, and stop checking the context.

</blockquote></details>

Let's stick to checking the context less often. Now I recall Raphael suggesting an occasional context check and a frequent flag check, so I'm moving to that now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Looking good, but let's polish that test some more.

---

Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/run_control_test.go, line 85 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8utI-jbp4KR7cpBLC:-Kq8utI-jbp4KR7cpBLD:bhrcv3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L85)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 	// A slower query and a faster query. Both of these will run in parallel, since they're RETURNING NOTHING.
> 	const query1 = "INSERT INTO nums (SELECT generate_series AS val FROM generate_series(1, 20000000000) WHERE generate_series = (-1)) RETURNING NOTHING;"
> ```

I don't know man like what if this fails way sooner because you're trying to insert all this data?
Let's rely exclusively on statement filters and such hooks. Or, block a query on an intent written by some other query.

---

*[pkg/sql/run_control_test.go, line 105 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8vYPP943Qln11JF61:-Kq8vYPP943Qln11JF62:b32vk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L105)):*
> ```Go
> 							// if query2
> 							if strings.Contains(stmt, "(2)") {
> 								// Assumption: query1 is still running. Query2 has finished, but still cancel it.
> ```

:|

---

*[pkg/sql/run_control_test.go, line 112 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8uF15KHlRrHfkSySr:-Kq8uF15KHlRrHfkSySs:b-10rdxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L112)):*
> ```Go
> 								go func() {
> 									sem <- struct{}{}
> 									const cancelQuery = "CANCEL QUERY (SELECT query_id FROM [SHOW CLUSTER QUERIES] WHERE node_id = 1 AND query LIKE '%INSERT INTO nums VALUES%')"
> ```

sick :)

---

*[pkg/sql/run_control_test.go, line 121 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8xXMknFX7GiW6lMSV:-Kq8xXMknFX7GiW6lMSW:bsdmex6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L121)):*
> ```Go
> 								}()
> 								<-sem
> 								time.Sleep(time.Second * 2)
> ```

what's this about? Whatever it is, it's gotta go :P

---

*[pkg/sql/run_control_test.go, line 131 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8wbKtQQDGBUO3bt2U:-Kq8wbKtQQDGBUO3bt2V:b-a7od1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L131)):*
> ```Go
> 
> 	conn1 = tc.ServerConn(0)
> 	conn2 = tc.ServerConn(1)
> ```

nit: rename this `cancelConn`

---

*[pkg/sql/run_control_test.go, line 138 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8vh2O4zgAfwCfHloq:-Kq8vh2O4zgAfwCfHlor:b40d48q) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L138)):*
> ```Go
> 	// should error out.
> 	_, err := conn1.Exec(sqlToRun)
> 	if err != nil && !strings.Contains(err.Error(), "query execution cancelled") {
> ```

there's a `testutils.IsError()` or smth like that. And also there's hopefully a pgcode we can test for.

But are you testing what you think you're testing? Who's to tell which one of the two queries encountered this error (cause the parallelizeQueue just returns a pretty arbitrary error, right?). I think we may need some new mechanism (`StatementFilter`) to inspect the result of a particular query.

---

*[pkg/sql/run_control_test.go, line 144 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8x64C1HKuYS8kJprP:-Kq8x64C1HKuYS8kJprQ:b2vm76z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L144)):*
> ```Go
> 	}
> 
> 	select {
> ```

I don't think this is necessary. If the cancel didn't run, you wouldn't get here.
Just wait on `errChan` (and always write on it, `nil` on success)

---

*[pkg/sql/session.go, line 212 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8y0lS4MvlfhHg5fFY:-Kq8y0lS4MvlfhHg5fFZ:b-64592o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/session.go#L212)):*
> ```Go
> // the same transaction (which could be running in parallel), then the shared
> // transaction context will be closed but isCancelled won't be set to true.
> func (q *queryMeta) isQueryCancelled(checkContext bool) bool {
> ```

nit: s/checkContext/checkTxn

the caller shouldn't know anything about any contexts.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-:-Kq8y7-zLTy28hqM0t7U:b-h3r8u2)*
<!-- Sent from Reviewable.io -->

itsbilal commented:
Added two new hooks to executor (since `StatementFilter` was insufficient for parallel test cases), and rewrote the test to use those.

---

Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/run_control_test.go, line 85 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8utI-jbp4KR7cpBLC:-Kq9nPZbelS0feVMXSDQ:bxadu0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L85)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't know man like what if this fails way sooner because you're trying to insert all this data?
Let's rely exclusively on statement filters and such hooks. Or, block a query on an intent written by some other query.
</blockquote></details>

Note the `WHERE` clause. It'll actually never insert anything, and spend a lot of time on the `SELECT`.  Either way, I've rewritten this test with newly-added executor hooks that makes it a lot more reliable.

---

*[pkg/sql/run_control_test.go, line 105 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8vYPP943Qln11JF61:-KqPHJZu6FJKPuT9x-o1:bl96j6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L105)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

:|
</blockquote></details>

Not an assumption anymore.

---

*[pkg/sql/run_control_test.go, line 121 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8xXMknFX7GiW6lMSV:-Kq9nOSDGHeKDgj-ynem:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L121)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what's this about? Whatever it is, it's gotta go :P
</blockquote></details>

Done.

---

*[pkg/sql/run_control_test.go, line 138 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8vh2O4zgAfwCfHloq:-KqPI33YUbdcqLjJjGA5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L138)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

there's a `testutils.IsError()` or smth like that. And also there's hopefully a pgcode we can test for.

But are you testing what you think you're testing? Who's to tell which one of the two queries encountered this error (cause the parallelizeQueue just returns a pretty arbitrary error, right?). I think we may need some new mechanism (`StatementFilter`) to inspect the result of a particular query.
</blockquote></details>

Done.

---

*[pkg/sql/run_control_test.go, line 144 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8x64C1HKuYS8kJprP:-Kq9r9VhdtAgOHakXtl0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/run_control_test.go#L144)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't think this is necessary. If the cancel didn't run, you wouldn't get here.
Just wait on `errChan` (and always write on it, `nil` on success)
</blockquote></details>

Done.

---

*[pkg/sql/session.go, line 212 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-Kq8y0lS4MvlfhHg5fFY:-KqPIDrd3wG_2Jnil9wg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4001e0c282cdab9bc4356fc481024ebe2c9b968/pkg/sql/session.go#L212)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: s/checkContext/checkTxn

the caller shouldn't know anything about any contexts.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-:-KqPIvC4T7sAluUoWo2K:b-i1cuip)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU4ywaKj66VuqsCKO_:-KqU4ywaKj66VuqsCKOa:bg8vgsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/executor.go#L309)):*
> ```Go
> 	// BeforeExecute is called by the Executor before plan execution. It is useful
> 	// for synchronizing statement execution, such as with parallel statemets.
> 	BeforeExecute func(ctx context.Context, stmt string, isDistributed bool)
> ```

the `isDistributed` param seems unnecessary to me. You're not passing anything else, like for example the plan, so why choose this arbitrary execution detail? If anything, I think you should pass a flag indicating whether the execution is parallelized - because that one tells you something about the state of the session as a whole - is it blocked or not?

---

*[pkg/sql/executor.go, line 313 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU4Mup0HQ4XkuGEzj9:-KqU4Mup0HQ4XkuGEzjA:bwwwctl) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/executor.go#L313)):*
> ```Go
> 	// AfterExecute is like StatementFilter, but it runs in the same goroutine of the
> 	// statement. Note that this function gets the actual execution error/result
> 	// which could differ from the result the client got (such as in the parallel case).
> ```

if the only case you have in mind in this comment is the parallel execution case, I wouldn't talk about it here; the results of parallel execution are a topic in itself, separate from a statement's result at this level.

---

*[pkg/sql/executor.go, line 314 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU6Qfrusj4ogDSgYK-:-KqU6Qfrusj4ogDSgYK0:bcjmh8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/executor.go#L314)):*
> ```Go
> 	// statement. Note that this function gets the actual execution error/result
> 	// which could differ from the result the client got (such as in the parallel case).
> 	AfterExecute func(ctx context.Context, stmt string, res *Result, err error)
> ```

A lot of times we make these hooks return an error so the test can inject an error. If you do that, take care about whether the code that calls the hook also has to override `res.Error` or not.

---

*[pkg/sql/run_control_test.go, line 118 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU7Oi6Xojb74vPf2oT:-KqU7Oi6Xojb74vPf2oU:b-eyv8lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/run_control_test.go#L118)):*
> ```Go
> 							if strings.Contains(stmt, "(1)") {
> 								// Ensure queryToBlock errored out with the cancellation error.
> 								if err == nil {
> ```

I suggest returning the errors from both queries on the channels and keeping the assertions in the main test body.

---

*[pkg/sql/run_control_test.go, line 159 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU7g9EW44gKLoBxBo3:-KqU7g9EW44gKLoBxBo4:b-11a32n) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/run_control_test.go#L159)):*
> ```Go
> 	// should error out.
> 	_, err := conn1.Exec(sqlToRun)
> 	if err != nil && !testutils.IsError(err, ".*query execution cancelled.*") {
> ```

if this doesn't already have an error code, I think we should give one (either find a suitable postgres one, or make a new crdb-specific one). And we should test code here rather than the message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-:-KqU7ayR1SS9WZy9wHz9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

itsbilal commented:
Review status: 0 of 5 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU4ywaKj66VuqsCKO_:-KqUHS29EDS5k9dArUJB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/executor.go#L309)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the `isDistributed` param seems unnecessary to me. You're not passing anything else, like for example the plan, so why choose this arbitrary execution detail? If anything, I think you should pass a flag indicating whether the execution is parallelized - because that one tells you something about the state of the session as a whole - is it blocked or not?
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 313 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU4Mup0HQ4XkuGEzj9:-KqUHRq9xJAS8LCdt90Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/executor.go#L313)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if the only case you have in mind in this comment is the parallel execution case, I wouldn't talk about it here; the results of parallel execution are a topic in itself, separate from a statement's result at this level.
</blockquote></details>

Done.

---

*[pkg/sql/run_control_test.go, line 118 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU7Oi6Xojb74vPf2oT:-KqUGlMFZYPETEeYkI60:b98m2z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/run_control_test.go#L118)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I suggest returning the errors from both queries on the channels and keeping the assertions in the main test body.
</blockquote></details>

The semantics around returning errors on the channel are different between the two queries - for instance, `queryToCancel` would need `err != nil` checks anyway before sending an error on the channel - or, the main goroutine would need a for loop to check the channel until it's closed. This setup puts the assertions closer to execution (and inside a query-specific if statement that already exists), so I expect it to be easier to understand for anyone reading the code.

---

*[pkg/sql/run_control_test.go, line 159 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU7g9EW44gKLoBxBo3:-KqUOZIlMOxcySfXVXTf:b-8dw54o) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/run_control_test.go#L159)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if this doesn't already have an error code, I think we should give one (either find a suitable postgres one, or make a new crdb-specific one). And we should test code here rather than the message.
</blockquote></details>

I wasn't able to make it work - for some reason the pgerror code isn't propagated all the way through to the `err` here. I'll investigate it and make a minor bugfix PR for just this change later - let's get the `txn` stuff in in the meantime.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/run_control_test.go, line 159 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-KqU7g9EW44gKLoBxBo3:-KqUjRkUBhPCa0ghk7qW:b-85yd1t) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b624a3d2f89f71ccefa7dca04a4c5aa954ed93/pkg/sql/run_control_test.go#L159)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I wasn't able to make it work - for some reason the pgerror code isn't propagated all the way through to the `err` here. I'll investigate it and make a minor bugfix PR for just this change later - let's get the `txn` stuff in in the meantime.
</blockquote></details>

ok... as long as you fix it :). I'd file an issue too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17219#-:-KqUjv5j04bDrdRvD0wO:bnfp4nl)*
<!-- Sent from Reviewable.io -->
